### PR TITLE
 Move warning before deepcopying the hyperparameters

### LINKED ISSUE #15131: Raise warning about saving nn.Module in save_hyperparameters earlier
## Proposed refactor

Raise warning about saving nn.Module in save_hyperparameters earlier

### Motivation

We have a warning when a user tries to "save" a nn.Module as hyperparameter here:

https://github.com/Lightning-AI/lightning/blob/8e83bfafcd92793910a76bbc41c58a91393da6fd/src/pytorch_lightning/utilities/parsing.py#L264-L267

This is to avoid redundantly big checkpoints and undesired behavior. The warning however is emitted after the deepcopy, and this can create memory issues or hangs and the warning will never be shown. 

### Pitch

Move the warning up so we can emit it earlier. 

### Additional context

Reported by user via social media.

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #15131

Suggestion welcome for a meaningful test

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda

### CODE REVIEW

**Review by carmocca — APPROVED** [review]
I don't think this needs a test.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by rohitgr7 — APPROVED** [review]

**Review by otaj — APPROVED** [review]
